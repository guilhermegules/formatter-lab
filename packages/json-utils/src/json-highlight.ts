export function highlightJSON(value: any, depth = 0): string {
  const indentClass = `depth-${depth}`;

  if (Array.isArray(value)) {
    const items = value
      .map(
        (item, index) =>
          `<div class="json-itens ${indentClass}">
            <span class="index">${index}:</span> ${highlightJSON(
              item,
              depth + 1
            )}
          </div>`
      )
      .join("");

    return `
      <span class="bracket">[</span>
      <div class="${indentClass} json-content">${items}</div>
      <span class="bracket">]</span>`;
  }

  if (value && typeof value === "object") {
    const entries = Object.entries(value)
      .map(
        ([key, item]) =>
          `<div class="json-itens ${indentClass}">
            <span class="key">"${key}"</span>: ${highlightJSON(item, depth + 1)}
          </div>`
      )
      .join("");

    return `
      <span class="bracket">{</span>
      <div class="${indentClass} json-content">${entries}</div>
      <span class="bracket">}</span>`;
  }

  if (value === null) {
    return `<span class="value">null</span>`;
  }

  if (typeof value === "number") {
    return `<span class="value">${value}</span>`;
  }

  if (typeof value === "string") {
    return `<span class="value">"${value}"</span>`;
  }

  return `<span class="value">${value}</span>`;
}

export function getJsonContainer(json: any) {
  return `<section class="json-container"><div class="json">${highlightJSON(
    json
  )}</div></section>`;
}
