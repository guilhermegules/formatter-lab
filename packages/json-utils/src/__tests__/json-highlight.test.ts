import { getJsonContainer, highlightJSON } from "../json-highlight";

describe("json highlight functions", () => {
  describe("highlightJSON", () => {
    it("should format a string value", () => {
      const result = highlightJSON("hello");
      expect(result).toBe('<span class="value">"hello"</span>');
    });

    it("should format a number value", () => {
      const result = highlightJSON(123);
      expect(result).toBe('<span class="value">123</span>');
    });

    it("should format a null value", () => {
      const result = highlightJSON(null);
      expect(result).toBe('<span class="value">null</span>');
    });

    it("should format a boolean values", () => {
      const result = highlightJSON(true);
      expect(result).toBe('<span class="value">true</span>');
    });

    it("should format an object", () => {
      const result = highlightJSON({ key: "value", num: 123 });
      expect(result).toMatchSnapshot();
    });

    it("should format nested objects and arrays", () => {
      const result = highlightJSON({
        obj: { key: "value" },
        arr: [1, 2, { nested: "deep" }],
      });
      expect(result).toMatchSnapshot();
    });

    it("should format nested objects and arrays", () => {
      const result = highlightJSON({
        obj: { key: "value" },
        arr: [1, 2, { nested: "deep" }],
      });
      expect(result).toMatch(/<span class="bracket">{<\/span>/);
      expect(result).toMatch(/<span class="key">"obj"<\/span>:/);
      expect(result).toMatch(/<span class="value">"value"<\/span>/);
      expect(result).toMatch(/<span class="key">"nested"<\/span>/);
      expect(result).toMatch(/<span class="bracket">\[<\/span>/);
    });
  });

  describe("getJsonContainer", () => {
    it("should wrap highlighted JSON in a container", () => {
      const json = { key: "value" };
      const result = getJsonContainer(json);
      expect(result).toContain('<section class="json-container">');
      expect(result).toContain('<div class="json">');
      expect(result).toContain(
        '<span class="key">"key"</span>: <span class="value">"value"</span>'
      );
    });

    it("should work with an empty object", () => {
      const json = {};
      const result = getJsonContainer(json);
      expect(result).toContain('<section class="json-container">');
      expect(result).toContain('<span class="bracket">{</span>');
      expect(result).toContain('<span class="bracket">}</span>');
    });

    it("should work with an empty array", () => {
      const json: any[] = [];
      const result = getJsonContainer(json);
      expect(result).toContain('<section class="json-container">');
      expect(result).toContain('<span class="bracket">[</span>');
      expect(result).toContain('<span class="bracket">]</span>');
    });
  });
});
