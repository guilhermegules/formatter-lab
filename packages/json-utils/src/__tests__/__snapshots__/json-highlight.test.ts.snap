// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json highlight functions highlightJSON should format an object 1`] = `
"
      <span class="bracket">{</span>
      <div class="depth-0 json-content"><div class="json-itens depth-0">
            <span class="key">"key"</span>: <span class="value">"value"</span>
          </div><div class="json-itens depth-0">
            <span class="key">"num"</span>: <span class="value">123</span>
          </div></div>
      <span class="bracket">}</span>"
`;

exports[`json highlight functions highlightJSON should format nested objects and arrays 1`] = `
"
      <span class="bracket">{</span>
      <div class="depth-0 json-content"><div class="json-itens depth-0">
            <span class="key">"obj"</span>: 
      <span class="bracket">{</span>
      <div class="depth-1 json-content"><div class="json-itens depth-1">
            <span class="key">"key"</span>: <span class="value">"value"</span>
          </div></div>
      <span class="bracket">}</span>
          </div><div class="json-itens depth-0">
            <span class="key">"arr"</span>: 
      <span class="bracket">[</span>
      <div class="depth-1 json-content"><div class="json-itens depth-1">
            <span class="index">0:</span> <span class="value">1</span>
          </div><div class="json-itens depth-1">
            <span class="index">1:</span> <span class="value">2</span>
          </div><div class="json-itens depth-1">
            <span class="index">2:</span> 
      <span class="bracket">{</span>
      <div class="depth-2 json-content"><div class="json-itens depth-2">
            <span class="key">"nested"</span>: <span class="value">"deep"</span>
          </div></div>
      <span class="bracket">}</span>
          </div></div>
      <span class="bracket">]</span>
          </div></div>
      <span class="bracket">}</span>"
`;
